File: bindings/wasm/src/quantum/post_quantum_algorithms.ts
import { Memory, Pointer, Result } from './types';
import { satox } from './satox';

export interface AlgorithmInfo {
    name: string;
    securityLevel: number;
    keySize: number;
    signatureSize: number;
    isRecommended: boolean;
    description: string;
}

export class PostQuantumAlgorithms {
    private algorithms: Pointer;
    private memory: Memory;

    constructor(memory: Memory) {
        this.memory = memory;
        this.algorithms = satox.post_quantum_algorithms_create();
        if (!this.algorithms) {
            throw new Error('Failed to create post-quantum algorithms');
        }
    }

    initialize(): void {
        const result = satox.post_quantum_algorithms_initialize(this.algorithms);
        if (result !== 0) {
            throw new Error('Failed to initialize post-quantum algorithms');
        }
    }

    shutdown(): void {
        const result = satox.post_quantum_algorithms_shutdown(this.algorithms);
        if (result !== 0) {
            throw new Error('Failed to shutdown post-quantum algorithms');
        }
    }

    generateKeyPair(algorithm: string): Result<{ publicKey: Uint8Array; privateKey: Uint8Array }> {
        const keyPair = satox.post_quantum_algorithms_generate_key_pair(
            this.algorithms,
            this.memory.writeString(algorithm)
        );

        if (!keyPair) {
            throw new Error('Failed to generate key pair');
        }

        const publicKey = this.memory.readBytes(keyPair.public_key, keyPair.public_key_len);
        const privateKey = this.memory.readBytes(keyPair.private_key, keyPair.private_key_len);

        this.memory.free(keyPair.public_key);
        this.memory.free(keyPair.private_key);

        return {
            publicKey,
            privateKey
        };
    }

    sign(algorithm: string, privateKey: Uint8Array, data: Uint8Array): Uint8Array {
        const signature = satox.post_quantum_algorithms_sign(
            this.algorithms,
            this.memory.writeString(algorithm),
            this.memory.writeBytes(privateKey),
            privateKey.length,
            this.memory.writeBytes(data),
            data.length
        );

        if (!signature) {
            throw new Error('Failed to sign data');
        }

        const result = this.memory.readBytes(signature.data, signature.length);
        this.memory.free(signature.data);

        return result;
    }

    verify(algorithm: string, publicKey: Uint8Array, data: Uint8Array, signature: Uint8Array): boolean {
        const result = satox.post_quantum_algorithms_verify(
            this.algorithms,
            this.memory.writeString(algorithm),
            this.memory.writeBytes(publicKey),
            publicKey.length,
            this.memory.writeBytes(data),
            data.length,
            this.memory.writeBytes(signature),
            signature.length
        );

        return result === 0;
    }

    encrypt(algorithm: string, publicKey: Uint8Array, data: Uint8Array): Uint8Array {
        const encrypted = satox.post_quantum_algorithms_encrypt(
            this.algorithms,
            this.memory.writeString(algorithm),
            this.memory.writeBytes(publicKey),
            publicKey.length,
            this.memory.writeBytes(data),
            data.length
        );

        if (!encrypted) {
            throw new Error('Failed to encrypt data');
        }

        const result = this.memory.readBytes(encrypted.data, encrypted.length);
        this.memory.free(encrypted.data);

        return result;
    }

    decrypt(algorithm: string, privateKey: Uint8Array, encrypted: Uint8Array): Uint8Array {
        const decrypted = satox.post_quantum_algorithms_decrypt(
            this.algorithms,
            this.memory.writeString(algorithm),
            this.memory.writeBytes(privateKey),
            privateKey.length,
            this.memory.writeBytes(encrypted),
            encrypted.length
        );

        if (!decrypted) {
            throw new Error('Failed to decrypt data');
        }

        const result = this.memory.readBytes(decrypted.data, decrypted.length);
        this.memory.free(decrypted.data);

        return result;
    }

    getAlgorithmInfo(algorithm: string): AlgorithmInfo {
        const info = satox.post_quantum_algorithms_get_algorithm_info(
            this.algorithms,
            this.memory.writeString(algorithm)
        );

        if (!info) {
            throw new Error('Failed to get algorithm info');
        }

        return {
            name: this.memory.readString(info.name),
            securityLevel: info.security_level,
            keySize: info.key_size,
            signatureSize: info.signature_size,
            isRecommended: info.is_recommended === 1,
            description: this.memory.readString(info.description)
        };
    }

    getAvailableAlgorithms(): string[] {
        const algorithms = satox.post_quantum_algorithms_get_available_algorithms(
            this.algorithms
        );

        if (!algorithms) {
            throw new Error('Failed to get available algorithms');
        }

        const result: string[] = [];
        for (let i = 0; i < algorithms.count; i++) {
            result.push(this.memory.readString(algorithms.list[i]));
        }

        return result;
    }

    isAlgorithmAvailable(algorithm: string): boolean {
        const result = satox.post_quantum_algorithms_is_algorithm_available(
            this.algorithms,
            this.memory.writeString(algorithm)
        );

        return result === 1;
    }

    isAlgorithmRecommended(algorithm: string): boolean {
        const result = satox.post_quantum_algorithms_is_algorithm_recommended(
            this.algorithms,
            this.memory.writeString(algorithm)
        );

        return result === 1;
    }

    getDefaultAlgorithm(): string {
        return satox.post_quantum_algorithms_get_default_algorithm(this.algorithms);
    }

    destroy(): void {
        if (this.algorithms) {
            satox.post_quantum_algorithms_destroy(this.algorithms);
            this.algorithms = null;
        }
    }
} 

File: bindings/python/satox/quantum/post_quantum_algorithms.py
"""
Post-Quantum Algorithms Python Bindings
"""

from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from satox_bindings.core.error_handling import SatoxError
from satox.quantum.native import satox

@dataclass
class AlgorithmInfo:
    """Information about a post-quantum algorithm."""
    name: str
    security_level: int
    key_size: int
    signature_size: int
    is_recommended: bool
    description: str

class PostQuantumAlgorithms:
    """Post-quantum cryptographic algorithms implementation."""

    def __init__(self):
        """Initialize a new PostQuantumAlgorithms instance."""
        self._algorithms = satox.PostQuantumAlgorithms()
        self._initialized = False

    def initialize(self) -> bool:
        """
        Initialize the post-quantum algorithms system.

        Returns:
            bool: True if initialization was successful

        Raises:
            SatoxError: If initialization fails
        """
        if not self._initialized:
            self._initialized = self._algorithms.initialize()
        return self._initialized

    def shutdown(self) -> bool:
        """
        Shutdown the post-quantum algorithms system.

        Returns:
            bool: True if shutdown was successful

        Raises:
            SatoxError: If shutdown fails
        """
        if self._initialized:
            self._initialized = not self._algorithms.shutdown()
        return not self._initialized

    def generate_key_pair(self, algorithm_name: str) -> Tuple[bytes, bytes]:
        """
        Generate a key pair using the specified algorithm.

        Args:
            algorithm_name (str): The name of the algorithm to use

        Returns:
            Tuple[bytes, bytes]: A tuple containing (public_key, private_key)

        Raises:
            SatoxError: If the system is not initialized or generation fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name:
            raise SatoxError("Invalid algorithm name", "INVALID_INPUT")
        return self._algorithms.generate_key_pair(algorithm_name)

    def sign(self, algorithm_name: str, private_key: bytes, data: bytes) -> bytes:
        """
        Sign data using the specified algorithm and private key.

        Args:
            algorithm_name (str): The name of the algorithm to use
            private_key (bytes): The private key to use for signing
            data (bytes): The data to sign

        Returns:
            bytes: The signature

        Raises:
            SatoxError: If the system is not initialized or signing fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name or not private_key or not data:
            raise SatoxError("Invalid input", "INVALID_INPUT")
        return self._algorithms.sign(algorithm_name, private_key, data)

    def verify(self, algorithm_name: str, public_key: bytes, data: bytes, signature: bytes) -> bool:
        """
        Verify a signature using the specified algorithm and public key.

        Args:
            algorithm_name (str): The name of the algorithm to use
            public_key (bytes): The public key to use for verification
            data (bytes): The data that was signed
            signature (bytes): The signature to verify

        Returns:
            bool: True if the signature is valid

        Raises:
            SatoxError: If the system is not initialized or verification fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name or not public_key or not data or not signature:
            raise SatoxError("Invalid input", "INVALID_INPUT")
        return self._algorithms.verify(algorithm_name, public_key, data, signature)

    def encrypt(self, algorithm_name: str, public_key: bytes, data: bytes) -> bytes:
        """
        Encrypt data using the specified algorithm and public key.

        Args:
            algorithm_name (str): The name of the algorithm to use
            public_key (bytes): The public key to use for encryption
            data (bytes): The data to encrypt

        Returns:
            bytes: The encrypted data

        Raises:
            SatoxError: If the system is not initialized or encryption fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name or not public_key or not data:
            raise SatoxError("Invalid input", "INVALID_INPUT")
        return self._algorithms.encrypt(algorithm_name, public_key, data)

    def decrypt(self, algorithm_name: str, private_key: bytes, encrypted_data: bytes) -> bytes:
        """
        Decrypt data using the specified algorithm and private key.

        Args:
            algorithm_name (str): The name of the algorithm to use
            private_key (bytes): The private key to use for decryption
            encrypted_data (bytes): The data to decrypt

        Returns:
            bytes: The decrypted data

        Raises:
            SatoxError: If the system is not initialized or decryption fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name or not private_key or not encrypted_data:
            raise SatoxError("Invalid input", "INVALID_INPUT")
        return self._algorithms.decrypt(algorithm_name, private_key, encrypted_data)

    def get_algorithm_info(self, algorithm_name: str) -> AlgorithmInfo:
        """
        Get information about a specific algorithm.

        Args:
            algorithm_name (str): The name of the algorithm

        Returns:
            AlgorithmInfo: Information about the algorithm

        Raises:
            SatoxError: If the system is not initialized or retrieval fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name:
            raise SatoxError("Invalid algorithm name", "INVALID_INPUT")
        return self._algorithms.get_algorithm_info(algorithm_name)

    def get_available_algorithms(self) -> List[str]:
        """
        Get a list of all available algorithms.

        Returns:
            List[str]: List of available algorithm names

        Raises:
            SatoxError: If the system is not initialized
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        return self._algorithms.get_available_algorithms()

    def get_recommended_algorithms(self) -> List[str]:
        """
        Get a list of recommended algorithms.

        Returns:
            List[str]: List of recommended algorithm names

        Raises:
            SatoxError: If the system is not initialized
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        return self._algorithms.get_recommended_algorithms()

    def is_algorithm_available(self, algorithm_name: str) -> bool:
        """
        Check if an algorithm is available.

        Args:
            algorithm_name (str): The name of the algorithm to check

        Returns:
            bool: True if the algorithm is available

        Raises:
            SatoxError: If the system is not initialized
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name:
            raise SatoxError("Invalid algorithm name", "INVALID_INPUT")
        return self._algorithms.is_algorithm_available(algorithm_name)

    def is_algorithm_recommended(self, algorithm_name: str) -> bool:
        """
        Check if an algorithm is recommended.

        Args:
            algorithm_name (str): The name of the algorithm to check

        Returns:
            bool: True if the algorithm is recommended

        Raises:
            SatoxError: If the system is not initialized
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name:
            raise SatoxError("Invalid algorithm name", "INVALID_INPUT")
        return self._algorithms.is_algorithm_recommended(algorithm_name)

    def get_algorithm_security_level(self, algorithm_name: str) -> int:
        """
        Get the security level of an algorithm.

        Args:
            algorithm_name (str): The name of the algorithm

        Returns:
            int: The security level

        Raises:
            SatoxError: If the system is not initialized or retrieval fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name:
            raise SatoxError("Invalid algorithm name", "INVALID_INPUT")
        return self._algorithms.get_algorithm_security_level(algorithm_name)

    def get_algorithm_key_size(self, algorithm_name: str) -> int:
        """
        Get the key size for an algorithm.

        Args:
            algorithm_name (str): The name of the algorithm

        Returns:
            int: The key size in bits

        Raises:
            SatoxError: If the system is not initialized or retrieval fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name:
            raise SatoxError("Invalid algorithm name", "INVALID_INPUT")
        return self._algorithms.get_algorithm_key_size(algorithm_name)

    def get_algorithm_signature_size(self, algorithm_name: str) -> int:
        """
        Get the signature size for an algorithm.

        Args:
            algorithm_name (str): The name of the algorithm

        Returns:
            int: The signature size in bytes

        Raises:
            SatoxError: If the system is not initialized or retrieval fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name:
            raise SatoxError("Invalid algorithm name", "INVALID_INPUT")
        return self._algorithms.get_algorithm_signature_size(algorithm_name)

    def is_initialized(self) -> bool:
        """
        Check if the post-quantum algorithms system is initialized.

        Returns:
            bool: True if initialized
        """
        return self._initialized

    def get_version(self) -> str:
        """
        Get the version of the post-quantum algorithms system.

        Returns:
            str: The version string
        """
        return self._algorithms.get_version()

    def get_default_algorithm(self) -> str:
        """
        Get the current default algorithm.

        Returns:
            str: The name of the default algorithm

        Raises:
            SatoxError: If the system is not initialized
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        return self._algorithms.get_default_algorithm()

    def set_default_algorithm(self, algorithm_name: str) -> bool:
        """
        Set the default algorithm.

        Args:
            algorithm_name (str): The name of the algorithm to set as default

        Returns:
            bool: True if setting was successful

        Raises:
            SatoxError: If the system is not initialized or setting fails
        """
        if not self._initialized:
            raise SatoxError("Post-Quantum Algorithms not initialized", "NOT_INITIALIZED")
        if not algorithm_name:
            raise SatoxError("Invalid algorithm name", "INVALID_INPUT")
        return self._algorithms.set_default_algorithm(algorithm_name) 